service/tbc/tbc.go

```diff
@@ -2656,10 +2683,37 @@ func (s *Server) synced(ctx context.Context) (si SyncInfo) {
 			Height: keystoneBH.Height,
 		}
 		si.Keystone = *keystoneHH
-		if keystoneHH.Hash.IsEqual(&bhb.Hash) {
+	}
+
+	if s.cfg.ZKIndex {
+		// Perform additional zk indexer tests.
+		zkBH, err := s.zki.IndexerAt(ctx)
+		if err != nil {
+			zkBH = &tbcd.BlockHeader{}
+		}
+		zkHH := &HashHeight{
+			Hash:   zkBH.Hash,
+			Height: zkBH.Height,
+		}
+		si.ZK = *zkHH
+	}
+
+	if utxoHH.Hash.IsEqual(&bhb.Hash) && txHH.Hash.IsEqual(&bhb.Hash) &&
+		!s.indexing && !blksMissing {
+		// If keystone and zk indexers are disabled we are synced.
+		if !s.cfg.HemiIndex && !s.cfg.ZKIndex {
 			si.Synced = true
 			return
 		}
+		if !(s.cfg.HemiIndex && si.Keystone.Hash.IsEqual(&bhb.Hash)) {
+			// Keystone index not synced
+			return
+		}
+		if !(s.cfg.ZKIndex && si.ZK.Hash.IsEqual(&bhb.Hash)) {
```

Comment:
De Morgan's law (usually I wouldn't comment, but here I think this is actually more readable):

```suggestion
		if !s.cfg.HemiIndex || !si.Keystone.Hash.IsEqual(&bhb.Hash) {
			// Keystone index not synced
			return
		}
		if !s.cfg.ZKIndex || !si.ZK.Hash.IsEqual(&bhb.Hash) {
```

Reply:
Never. !() always wins over mixing ! ||